### PR TITLE
add ability to run scripts directly (without processors) within Ingest Pipelines

### PR DESCRIPTION
**Pull request by talevy**
- add `script` field to pipelines as a replacement to executing processors
- introduce a specific IngestContext script context for executing within pipelines

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java [hunk 1]
```diff
@@ -340,4 +346,26 @@ public static Processor readProcessor(Map<String, Processor.Factory> processorFa
         }
         throw newConfigurationException(type, tag, null, "No processor type exists with name [" + type + "]");
     }
+
+    @SuppressWarnings("unchecked")
+    static Script readOptionalScript(String processorType, String processorTag, Map<String, Object> configuration,
+                                            String propertyName) {
+        try {
+            XContentBuilder builder = XContentBuilder.builder(JsonXContent.jsonXContent);
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: Can we create the XContentBuilder after the scriptObject == null condition?

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java [hunk 2]
```diff
@@ -340,4 +346,26 @@ public static Processor readProcessor(Map<String, Processor.Factory> processorFa
...
+            Object scriptObject = configuration.remove(propertyName);
+            if (scriptObject == null) {
+                return null;
+            }
+            if (scriptObject instanceof String) {
+                builder.value(scriptObject);
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: I think that we can save the instanceof checks, builder.value(Object) does it already

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java [hunk 3]
```diff
@@ -340,4 +346,26 @@ public static Processor readProcessor(Map<String, Processor.Factory> processorFa
...
+            } else {
+                builder.map((Map<String, Object>) scriptObject);
+            }
+            JsonXContentParser parser = new JsonXContentParser(NamedXContentRegistry.EMPTY,
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: Is there a reason why we don't use XContentHelper.createParser(registry, bytes,contentType) ?

Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: no specific good reason, just still getting used to the more idiomatic helper libraries for working with xcontent

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java (hunk 2 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: ah, didn't bother to check. perfect!

File: core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java (hunk 1 above)
Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/ConfigurationUtils.java]: yeah. that would be better, huh.

**Comment by talevy** [comment]
thanks for the quick scan @tlrx!

I mainly opened this PR to see if this is even something we are interested in doing.
I think it would be really nice to be able to leverage Ingest Processors from within Painless scripts — therefore removing the need for a json-defined pipeline. The main blocker for this is painless whitelisting, which should be coming soon

**Comment by dakrone** [comment]
@talevy is this still applicable? I know it's a WIP PR, but it has some conflicts now, so can it be closed?

**Comment by talevy** [comment]
@dakrone sure, I'll close it.

I left it open since I do intend on merging this type of feature at some point.
I punted since this level of directness would only be stronger once Painless adds whitelisting.

**Pull request closed without merging**